packages/backend-lib/src/destinations/amazonses.ts

```diff
@@ -0,0 +1,199 @@
+import Https from "node:https";
+
+import {
+  MessageTag,
+  SendEmailCommand,
+  type SendEmailCommandInput,
+  SendEmailCommandOutput,
+  SESClient,
+  SESServiceException,
+} from "@aws-sdk/client-ses";
+import { err, Result, ResultAsync } from "neverthrow";
+import SnsPayloadValidator from "sns-payload-validator";
+
+import { submitBatch } from "../apps";
+import logger from "../logger";
+import {
+  AmazonSesConfig,
+  AmazonSesEventPayload,
+  AmazonSesMailFields,
+  AmazonSesNotificationType,
+  AmazonSNSEvent,
+  AmazonSNSSubscriptionEvent,
+  AmazonSNSUnsubscribeEvent,
+  BatchTrackData,
+  EventType,
+  InternalEventType,
+} from "../types";
+
+function unwrapTag(tagName: string, tags: Record<string, string[]>) {
+  if (!tags[tagName]) {
+    return null;
+  }
+  // disable the nonnull assertion eslint rule
+  // eslint-disable-next-line
+  const tag = tags[tagName]!;
+
+  return tag[0];
+}
+
+// README the typescript types on this are wrong, body is not of type string,
+// it's a parsed JSON object
+function guardResponseError(e: unknown): SESServiceException {
+  if (e instanceof SESServiceException) {
+    return e;
+  }
+  throw e;
+}
+
+export async function sendMail({
+  config,
+  mailData,
+}: {
+  config: AmazonSesConfig;
+  mailData: AmazonSesMailFields;
+}): Promise<Result<SendEmailCommandOutput, SESServiceException>> {
+  const { accessKeyId, secretAccessKey, region } = config;
+  const client = new SESClient({
+    region,
+    credentials: {
+      accessKeyId,
+      secretAccessKey,
+    },
+  });
+
+  const replyTo = mailData.replyTo
+    ? {
+        ReplyToAddresses: [mailData.replyTo],
+      }
+    : {};
+
+  const tags = mailData.tags
+    ? {
+        Tags: Object.keys(mailData.tags).reduce(
+          (a: MessageTag[], k: string) => {
+            return mailData.tags
+              ? [{ Name: k, Value: mailData.tags[k] }, ...a]
+              : a;
+          },
+          [],
+        ),
+      }
+    : {};
+
+  const input: SendEmailCommandInput = {
+    Source: mailData.from,
+    Destination: {
+      ToAddresses: [mailData.to],
+    },
+    Message: {
+      Subject: {
+        Data: mailData.subject,
+        Charset: "UTF-8",
+      },
+      Body: {
+        Html: {
+          Data: mailData.html,
+          Charset: "UTF-8",
+        },
+      },
+    },
+    ...tags,
+    ...replyTo,
+  };
+
+  const command = new SendEmailCommand(input);
+
+  return ResultAsync.fromPromise(client.send(command), guardResponseError).map(
+    (resultArray) => resultArray,
+  );
+}
+
+export async function submitAmazonSesEvents(
+  event: AmazonSesEventPayload,
+): Promise<ResultAsync<void, Error>> {
+  // TODO: Amazon may batch requests (if we send with multiple To: addresses? or with the BatchTemplated endpoint).  We should map over the receipients.
+  logger().debug(event);
+
+  const workspaceId = unwrapTag("workspaceId", event.mail.tags) ?? null;
```

Comment:
Is it necessary to do `?? null` it looks like this function returns `string | null` already.